demos/evp/aesgcm.c

```diff
@@ -47,75 +48,147 @@ static const unsigned char gcm_tag[] = {
     0x98, 0xf7, 0x7e, 0x0c
 };
 
-void aes_gcm_encrypt(void)
+int aes_gcm_encrypt(void)
 {
+    int ret = 0;
     EVP_CIPHER_CTX *ctx;
+    EVP_CIPHER *cipher;
     int outlen, tmplen;
     unsigned char outbuf[1024];
+
     printf("AES GCM Encrypt:\n");
     printf("Plaintext:\n");
     BIO_dump_fp(stdout, gcm_pt, sizeof(gcm_pt));
-    ctx = EVP_CIPHER_CTX_new();
-    /* Set cipher type and mode */
-    EVP_EncryptInit_ex(ctx, EVP_aes_256_gcm(), NULL, NULL, NULL);
+
+    /* Create a context for the encrypt operation */
+    if ((ctx = EVP_CIPHER_CTX_new()) == NULL)
+        goto err;
+
+    /* Fetch the cipher implementation */
+    if ((cipher = EVP_CIPHER_fetch(NULL, "AES-256-GCM", NULL)) == NULL)
+        goto err;
+
+    /* Initialise encrypt operation with the cipher & mode */
+    if (!EVP_EncryptInit_ex(ctx, cipher, NULL, NULL, NULL))
+        goto err;
+
     /* Set IV length if default 96 bits is not appropriate */
-    EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_AEAD_SET_IVLEN, sizeof(gcm_iv), NULL);
+    if (!EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_AEAD_SET_IVLEN, sizeof(gcm_iv), NULL))
+        goto err;
+
     /* Initialise key and IV */
-    EVP_EncryptInit_ex(ctx, NULL, NULL, gcm_key, gcm_iv);
+    if (!EVP_EncryptInit_ex(ctx, NULL, NULL, gcm_key, gcm_iv))
+        goto err;
+
     /* Zero or more calls to specify any AAD */
-    EVP_EncryptUpdate(ctx, NULL, &outlen, gcm_aad, sizeof(gcm_aad));
+    if (!EVP_EncryptUpdate(ctx, NULL, &outlen, gcm_aad, sizeof(gcm_aad)))
+        goto err;
+
     /* Encrypt plaintext */
-    EVP_EncryptUpdate(ctx, outbuf, &outlen, gcm_pt, sizeof(gcm_pt));
+    if (!EVP_EncryptUpdate(ctx, outbuf, &outlen, gcm_pt, sizeof(gcm_pt)))
+        goto err;
+
     /* Output encrypted block */
     printf("Ciphertext:\n");
     BIO_dump_fp(stdout, outbuf, outlen);
+
     /* Finalise: note get no output for GCM */
-    EVP_EncryptFinal_ex(ctx, outbuf, &outlen);
+    if (!EVP_EncryptFinal_ex(ctx, outbuf, &outlen))
+        goto err;
+
     /* Get tag */
-    EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_AEAD_GET_TAG, 16, outbuf);
+    if (!EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_AEAD_GET_TAG, 16, outbuf))
```

Comment:
It is a bit weird to trash the outbuf here.. It did contain the encrypted data, before the tag wrote over the top of it..
outbuf + len would be the normal thing here.. or a seperate buffer for the tag..